samcli/lib/pipeline/bootstrap/stage_resources.yaml

```diff
@@ -0,0 +1,359 @@
+                  - MissingPipelineExecutionRole
+                  - !GetAtt PipelineExecutionRole.Arn
+                  - !Ref PipelineExecutionRoleArn
+                - Fn::If:
+                  - MissingCloudFormationExecutionRole
+                  - !GetAtt CloudFormationExecutionRole.Arn
+                  - !Ref CloudFormationExecutionRoleArn
+
+  PipelineExecutionRolePermissionPolicy:
+    Type: AWS::IAM::Policy
+    Condition: MissingPipelineExecutionRole
+    Properties:
+      PolicyName: PipelineExecutionRolePermissions
+      PolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Effect: Allow
+            Action: 'iam:PassRole'
+            Resource:
+              Fn::If:
+                - MissingCloudFormationExecutionRole
+                - !GetAtt CloudFormationExecutionRole.Arn
+                - !Ref CloudFormationExecutionRoleArn
+          - Effect: Allow
+            Action:
+              - "cloudformation:CreateChangeSet"
+              - "cloudformation:DescribeChangeSet"
+              - "cloudformation:ExecuteChangeSet"
+              - "cloudformation:DescribeStackEvents"
+              - "cloudformation:DescribeStacks"
+              - "cloudformation:GetTemplateSummary"
+              - "cloudformation:DescribeStackResource"
+            Resource: '*'
+          - Effect: Allow
+            Action:
+              - 's3:GetObject*'
+              - 's3:PutObject*'
+              - 's3:GetBucket*'
+              - 's3:List*'
+            Resource:
+              Fn::If:
+                - MissingArtifactsBucket
+                - - !Join [ '',[ !GetAtt ArtifactsBucket.Arn, '/*' ] ]
+                  - !GetAtt ArtifactsBucket.Arn
+                - - !Join [ '',[ !Ref ArtifactsBucketArn, '/*' ] ]
+                  - !Ref ArtifactsBucketArn
+          - Fn::If:
+            - MissingArtifactsBucket
+            - Effect: "Allow"
+              Action:
+                - "kms:Decrypt"
+                - "kms:DescribeKey"
+              Resource:
+                - !GetAtt ArtifactsBucketKey.Arn
+            - !Ref AWS::NoValue
+          - Fn::If:
+            - EcrRepoRequired
+            - Effect: "Allow"
+              Action: "ecr:GetAuthorizationToken"
+              Resource: "*"
+            - !Ref AWS::NoValue
+          - Fn::If:
+            - EcrRepoRequired
+            - Effect: "Allow"
+              Action:
+                - "ecr:GetDownloadUrlForLayer"
+                - "ecr:BatchGetImage"
+                - "ecr:BatchCheckLayerAvailability"
+                - "ecr:PutImage"
+                - "ecr:InitiateLayerUpload"
+                - "ecr:UploadLayerPart"
+                - "ecr:CompleteLayerUpload"
+              Resource:
+                Fn::If:
+                  - MissingEcrRepo
+                  - !GetAtt EcrRepo.Arn
+                  - !Ref EcrRepoArn
+            - !Ref AWS::NoValue
+      Roles:
+        - !Ref PipelineExecutionRole
+
+  EcrRepo:
```

Comment:
https://github.com/aws/aws-sam-cli/pull/2675 creates an ECR repo such that its per function. @CoshUS can you add your comments here? My concern is we do conflicting things via ci/cd vs guided deploy experience.

Reply:
We had a discussion on the overlapping parts of these two features. They have enough differences between them (command executor permission, image per function, and stack per account) that keeping them separate is the current option we are going for. However, we will make sure the ECR repo permissions are consistent and the documentation states clearly which one should be used for what use-case.

Reply:
That is a good call out. We will make sure of the consistency between the two functionality. But both are needed as they have different use-cases; in the non CI/CD case, `sam deploy` is ran by the user who has an administrative permissions to the account while in case of CI/CD `sam deploy` is ran through an IAM entity that has limited permissions. So, creating the ECR repo should be part of the `sam pipeline bootstrap` which is ran by the normal user(the one with the administrative access)